### PR TITLE
Context and functional components

### PR DESCRIPTION
**Pull request by eastcoastcoder**
I wanted to rewrite everything without the Redux dependency (using Context instead), with functional components that use hooks, and with async methods. This mostly works, in my limited testing I haven't ran into any blatantly obvious issues.

Though this really deviates from the project quite a bit especially with a heavy amount of linting to keep it as close to vanilla Airbnb style as possible. If this strays too far from the original project I'd be happy to maintain it in my own fork or adapt the code if appropriate.

### CODE REVIEW

**Comment by banshiAnton** [comment]
Hi @ethanx94, thank you for your PR, out team will review and test your PR

**Comment by stephanoparaskeva** [comment]
Can you do this for RNVideoChat also?

**Comment by ethanx94** [comment]
@stephanoparaskeva By request I've added RNVideoChat to get it up to speed with async functions, etc. This project had no real Redux dependency but I changed all the class based components to be functional at least. This one however, is even less tested as I do not have two phones on me currently to do so. Please run with the provided .vscode debug scripts (attach to packager after you have it running on a device) and see if you can track down any bugs.

File: RNVideoChat/src/components/VideoScreen/index.js [hunk 1]
```diff
@@ -1,258 +1,240 @@
...
-  }
-
-  componentWillUnmount() {
-    CallService.stopCall();
-    AuthService.logout();
-  }
-
-  componentDidUpdate(prevProps, prevState) {
+  useEffect(() => {
+    /*
```
Comment (ethanx94) [inline RNVideoChat/src/components/VideoScreen/index.js:40]: Forgot about this part...

**Comment by stephanoparaskeva** [comment]
> @stephanoparaskeva By request I've added RNVideoChat to get it up to speed with async functions, etc. This project had no real Redux dependency but I changed all the class based components to be functional at least. This one however, is even less tested as I do not have two phones on me currently to do so. Please run with the provided .vscode debug scripts (attach to packager after you have it running on a device) and see if you can track down any bugs.

Awesome stuff. Will test asap and fingers crossed

**Comment by stephanoparaskeva** [comment]
> @stephanoparaskeva By request I've added RNVideoChat to get it up to speed with async functions, etc. This project had no real Redux dependency but I changed all the class based components to be functional at least. This one however, is even less tested as I do not have two phones on me currently to do so. Please run with the provided .vscode debug scripts (attach to packager after you have it running on a device) and see if you can track down any bugs.

After testing with 2 mobiles the video calling seems to have only a few issues. 

* When you start a call, the speaker doesn't seem to be on. So you can only hear each device if you put the phone to your ear.
* The video stream of the remote device does not show on screen. So you basically see a fullscreen video of yourself on each phone as if only the local video is shown on each.

I can try and look further to fix these and will let you know but as of yesterday I haven't been able to locate what is causing this problem

**Comment by ethanx94** [comment]
Thanks for the updateRemoteStream fix. I noticed hangup doesn't bring you back. Since this branch was a few months ago I am getting further away from master. I need to sync back up and see if there's anything new I missed. I pushed up another update which might fix the cameras not working. I wasn't referencing the ref properly for "isAvailableToSwitch" which I think controls the swap camera button.

I do call the setSpeakerphoneOn in a useEffect any time the remoteStreams change...the value should be true. But you're right I'm not hearing anything on speaker at any point.

**Comment by stephanoparaskeva** [comment]
> Thanks for the updateRemoteStream fix. I noticed hangup doesn't bring you back. Since this branch was a few months ago I am getting further away from master. I need to sync back up and see if there's anything new I missed. I pushed up another update which might fix the cameras not working. I wasn't referencing the ref properly for "isAvailableToSwitch" which I think controls the swap camera button.
> 
> I do call the setSpeakerphoneOn in a useEffect any time the remoteStreams change...the value should be true. But you're right I'm not hearing anything on speaker at any point.

Have been testing and trying to debug but have had no luck. I can't seem to figure out how to fix these and offer consistent calling. Right now the problems I've found:

* Speaker doesn't work on IOS
* Android camera switch freezes 
* Calling and hanging up sometimes doesn't return to home screen
* [CRITICAL] Calling another phone and then hanging up. If you call again, it says 'User B is Busy' and call sound occurs on other phone but no prompt occurs. Have to refresh app to fix both phones.

Will have to work through these but I can't seem to get this to work consistently. Perhaps we've missed something. What's your thoughts?

**Comment by ethanx94** [comment]
Hey, sorry been a bit pressed with work this week. As far as the critical issue, I would check closely into the listeners. Its almost as though the state isn't resetting when the call ends. I tried to use a ref to keep the previous state and I observe the stateful value in my useEffect dependency array. The problem may very well lie there.

With that said my conversion to hooks may not be the most performant, and that should come as a disclaimer. There may be times when it is appropriate to do useCallback/useMemo but I'm not familiar enough with the codebase and hooks in general to make those optimizations just yet. I didn't personably run into this, but I know it might be a problem in the RNChat example with large dialogs.

Here is possibly the problem area in RNVideoChat.
image

**Comment by stephanoparaskeva** [comment]
> Hey, sorry been a bit pressed with work this week. As far as the critical issue, I would check closely into the listeners. Its almost as though the state isn't resetting when the call ends. I tried to use a ref to keep the previous state and I observe the stateful value in my useEffect dependency array. The problem may very well lie there.
> 
> With that said my conversion to hooks may not be the most performant, and that should come as a disclaimer. There may be times when it is appropriate to do useCallback/useMemo but I'm not familiar enough with the codebase and hooks in general to make those optimizations just yet. I didn't personably run into this, but I know it might be a problem in the RNChat example with large dialogs.
> 
> Here is possibly the problem area in RNVideoChat.
> image

It's awesome work. I've managed to fix these issues:
* Speaker phone doesn't work.
* Call doesn't return to home.
* Sometimes is busy is on.

The problem is that I can't make a PR into your work because my code is all TypeScript and also I completely removed the call_service and moved it all into video_screen and I use (react-native-callkeep)[https://github.com/react-native-webrtc/react-native-callkeep] instead of awesome alerts for the call screen And I made calling only peer-to-peer no conferences.
But to fix speaker phone you have to invoke incallmanager like this:
```JavaScript
  const onRemoteStreamListener = async (
    _session: Session,
    userId: number,
    stream: Stream
  ) => {
    InCallManager.start({ media: 'video' });
    setRemoteStreams([...remoteStreams, { userId, stream }]);
  };
```

The whole file is here:
https://pastebin.com/BR2NCzxY

### Problems I have now:
* Sometimes if I leave apps running for 20+ minutes they no longer respond to calls from eachother.
* Android phone doesn't work and is buggy so must fix this
* Switch camera no longer works (will be an easy fix)


Is it possible for you to help advise me on this one:
`Sometimes if I leave apps running for 20+ minutes they no longer respond to calls from eachother.` as I need to make calling consistent and I have no idea why this might not be working.

**Comment by ethanx94** [comment]
As far as the 20 minute idle time, there was a similar issue a while back. Maybe this is related? I would try some workarounds in this thread.
https://github.com/ConnectyCube/connectycube-reactnative-samples/issues/64

**Comment by stephanoparaskeva** [comment]
> As far as the 20 minute idle time, there was a similar issue a while back. Maybe this is related? I would try some workarounds in this thread.
> #64

After upgrading to 3.1.0 I think it may have fixed this for me.
I've integrated RN-CallKeep and here's the code. I've got lot's of issues right now and I'm working through them. It's really quite tricky:

```JavaScript
import React, { useRef, useState, useEffect } from 'react';
import {
  SafeAreaView,
  StatusBar,
  Platform,
  View,
  TouchableOpacity,
  StyleSheet,
} from 'react-native';
import ConnectyCube from 'react-native-connectycube';
import InCallManager from 'react-native-incall-manager';
import RNCallKeep from 'react-native-callkeep';
import MaterialIcon from 'react-native-vector-icons/MaterialIcons';
import { v4 as uuid } from 'uuid';
import VideoScreen from './video_screen';
import UsersSelect from './users_select';
import {
  Session,
  Stream,
  Listener,
  users,
  Extension,
  Track,
  MEDIA_OPTIONS,
  SETUP_OPTIONS,
} from '../config';
import useStateCallback from '../../components/common/hooks/state_callback';

const isIOS = Platform.OS === 'ios';

const CallControl = ({ route }: { route: any }) => {
  const initSession = {
    ID: '',
    initiatorID: 0,
    currentUserID: 0,
    opponentsIDs: [],
    getUserMedia: () => {},
    accept: () => {},
    call: () => {},
    stop: () => {},
    mute: () => {},
    unmute: () => {},
    reject: () => {},
  };
  const initExtension = { busy: false };

  const sess = useRef<Session>(initSession);
  const ext = useRef<Extension>(initExtension);
  const callId = useRef<string>('');
  const currentUser = route?.params?.currentUser || {};

  const [localStream, setLocalStreams] = useState<any>([]);
  const [remoteStreams, setRemoteStreams] = useStateCallback<
    [] | Array<Stream>
  >([]);

  const [selectedUserId, setSelectedUser] = useState<null | number>(null);

  const [isAudioMuted, setIsAudioMuted] = useState(false);
  const [isFrontCamera, setIsFrontCamera] = useState(true);

  const cam = true;

  const reset = async () => {
    const id = oneTimeId();
    RNCallKeep.endCall(id);
    await ConnectyCube.videochat.clearSession(sess.current.ID);
    RNCallKeep.endAllCalls();

    InCallManager.stop();
    setRemoteStreams([]);
    setLocalStreams([]);
    setSelectedUser(null);
    setIsAudioMuted(false);
    setIsFrontCamera(true);
    sess.current = initSession;
    ext.current = initExtension;
    callId.current = '';
  };

  const oneTimeId = () => {
    callId.current = callId.current || uuid();
    return callId.current;
  };

  // Start/Stop onPress call handlers
  const startCall = async () => {
    if (!selectedUserId) {
      return;
    }

    const t = ConnectyCube.videochat.CallType.VIDEO; // AUDIO is also possible
    sess.current = ConnectyCube.videochat.createNewSession([selectedUserId], t);

    const stream = await sess.current?.getUserMedia?.(MEDIA_OPTIONS);
    await sess.current?.call({});

    setLocalStreams([{ userId: 'localStream', stream }]);
    setRemoteStreams([{ userId: selectedUserId, stream: null }]);

    if (isIOS) {
      const id = oneTimeId();
      const name = users.find(u => u.id === selectedUserId)?.name || '';
      RNCallKeep.startCall(id, name, name, 'generic', cam);
    }
  };

  const stopCall = async () => {
    await sess.current?.stop?.({});
    await reset();
  };

  // CallKeep button actions
  const acceptCall = async () => {
    const id = oneTimeId();
    RNCallKeep.setCurrentCallActive(id);

    const stream = await sess.current?.getUserMedia?.(MEDIA_OPTIONS);
    await sess.current?.accept?.({});

    setLocalStreams([{ userId: 'localStream', stream }]);
    setRemoteStreams([{ userId: selectedUserId, stream: null }]);
  };

  const rejectCall = async () => {
    await sess.current?.reject?.(ext.current);
    await reset();
  };

  // ConnectyCall Listeners
  const onCallListener: Listener = async (session, _userId, extension) => {
    ext.current = extension || initExtension;
    sess.current = session || initSession;

    const name =
      users.find(u => u.id === sess?.current?.initiatorID)?.name || '';
    const id = oneTimeId();
    RNCallKeep.displayIncomingCall(id, name, name, 'generic', cam);
  };

  const onAcceptCallListener: Listener = async (session, _, extension) => {
    ext.current = extension || initExtension;
    sess.current = session || initSession;
  };

  const onRemoteStreamListener = async (
    _session: Session,
    userId: number,
    stream: Stream
  ) =>
    setRemoteStreams([...remoteStreams, { userId, stream }], () => {
      InCallManager.start({ media: 'video', auto: false });
      InCallManager.setSpeakerphoneOn(true);
      RNCallKeep.backToForeground();
    });

  const onRejectCallListener = async () => {
    await reset();
  };

  const onStopCallListener = async () => {
    await reset();
  };

  const onUserNotAnswerListener = async () => {
    if (!remoteStreams[0]) {
      await reset();
    }
  };

  // Init
  const initializeConnectyCube = () => {
    ConnectyCube.videochat.onCallListener = onCallListener;
    ConnectyCube.videochat.onAcceptCallListener = onAcceptCallListener;
    ConnectyCube.videochat.onRemoteStreamListener = onRemoteStreamListener;
    ConnectyCube.videochat.onRejectCallListener = onRejectCallListener;
    ConnectyCube.videochat.onStopCallListener = onStopCallListener;
    ConnectyCube.videochat.onUserNotAnswerListener = onUserNotAnswerListener;
  };

  const initializeCallKeep = async () => {
    await RNCallKeep.setup(SETUP_OPTIONS);
    RNCallKeep.setAvailable(true);
    RNCallKeep.addEventListener('answerCall', acceptCall);
    RNCallKeep.addEventListener('endCall', rejectCall);
  };

  const cleanup = () => {
    RNCallKeep.removeEventListener('answerCall');
    RNCallKeep.removeEventListener('endCall');
    stopCall();
  };

  useEffect(() => {
    initializeConnectyCube();
    initializeCallKeep();
    return cleanup;
  }, []);

  // Actions
  const toggleMute = () => {
    const mute = !isAudioMuted;
    setIsAudioMuted(mute);
    mute ? sess.current?.mute?.('audio') : sess.current?.unmute?.('audio');
  };

  const toggleCamera = () => {
    setIsFrontCamera(!isFrontCamera);
    localStream[0]?.stream
      ?.getVideoTracks?.()
      .forEach((track: Track) => track._switchCamera());
  };

  // Buttons
  const renderCallStartStopButton = () => (
    <TouchableOpacity
      style={[
        styles.buttonContainer,
        remoteStreams[0] ? styles.buttonCallEnd : styles.buttonCall,
      ]}
      onPress={remoteStreams[0] ? stopCall : startCall}>
      <MaterialIcon
        name={remoteStreams[0] ? 'call-end' : 'call'}
        size={32}
        color='white'
      />
    </TouchableOpacity>
  );

  const renderMuteButton = () => (
    <TouchableOpacity
      style={[styles.buttonContainer, styles.buttonMute]}
      onPress={toggleMute}>
      <MaterialIcon
        name={isAudioMuted ? 'mic-off' : 'mic'}
        size={32}
        color='white'
      />
    </TouchableOpacity>
  );

  const renderSwitchVideoSourceButton = () => (
    <TouchableOpacity
      style={[styles.buttonContainer, styles.buttonSwitch]}
      onPress={toggleCamera}>
      <MaterialIcon
        name={isFrontCamera ? 'camera-rear' : 'camera-front'}
        size={32}
        color='white'
      />
    </TouchableOpacity>
  );

  return (
    <SafeAreaView style={{ flex: 1, backgroundColor: 'black' }}>
      <StatusBar backgroundColor='black' barStyle='light-content' />
      {remoteStreams[0] ? (
        <View style={styles.inColumn}>
          <VideoScreen
            mirror={false}
            stream={remoteStreams?.[0]?.stream}
            userId={remoteStreams?.[0]?.userId}
          />
          <VideoScreen
            mirror={isFrontCamera}
            stream={localStream?.[0]?.stream}
            userId={localStream?.[0]?.userId}
          />
        </View>
      ) : (
        <UsersSelect
          currentUser={currentUser}
          selectedUserId={selectedUserId}
          setSelectedUser={setSelectedUser}
        />
      )}
      <SafeAreaView style={styles.container}>
        <View style={styles.toolBarItem}>
          {remoteStreams[0] && renderMuteButton()}
        </View>
        {(selectedUserId || remoteStreams[0]) && (
          <View style={styles.toolBarItem}>{renderCallStartStopButton()}</View>
        )}
        <View style={styles.toolBarItem}>
          {remoteStreams[0] && localStream && renderSwitchVideoSourceButton()}
        </View>
      </SafeAreaView>
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  container: {
    position: 'absolute',
    bottom: 20,
    left: 0,
    right: 0,
    height: 60,
    flex: 1,
    justifyContent: 'center',
    flexDirection: 'row',
    zIndex: 100,
  },
  toolBarItem: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
  buttonContainer: {
    height: 50,
    width: 50,
    borderRadius: 25,
    marginHorizontal: 25,
    justifyContent: 'center',
    alignItems: 'center',
  },
  buttonCall: {
    backgroundColor: 'green',
  },
  buttonCallEnd: {
    backgroundColor: 'red',
  },
  buttonMute: {
    backgroundColor: 'blue',
  },
  buttonSwitch: {
    backgroundColor: 'orange',
  },
  inColumn: {
    flex: 1,
    flexDirection: 'column',
  },
});

export default CallControl;```

**Comment by DaveLomber** [comment]
RNChat is already re-written in functional components and merged into master.
RNVideoChat - will be done with https://github.com/ConnectyCube/connectycube-reactnative-samples

We decide to keep using redux because of its popularity among developers.

**Comment by DaveLomber** [comment]
both RNChat and RNVideoChat are re-written with functional components, 
please refer to master branch for the latest src code

**Pull request closed without merging**